[CK_TILE] Enhance RMSNorm Accuracy: New Pipeline Pass for Selectable Implementation

MHYangAMD opened the pull request:
## Proposed changes

There are precision gaps between the [vllm hip implementation](https://github.com/vllm-project/vllm/blob/main/csrc/layernorm_kernels.cu) and CK when testing some models.

The results with this patch are much more closer to vllm hip, but still have some differences.

Example reduction output:
CK: float = 26543.632812, raw = 46cf5f44
vllm: float = 26543.630859, raw = 46cf5f43

The 1 bit difference occurs right after the reduction, and such 1 bit difference may cause incorrect rmsnorm results up to the level of 1e-2.


We suspect that the issue relates to the reduction method used in the native VLLM implementation (rocprim. [block reduce](https://github.com/ROCm/rocm-libraries/blob/develop/projects/rocprim/rocprim/include/rocprim/block/detail/block_reduce_warp_reduce.hpp#L222) [warp-reduce](https://github.com/ROCm/rocm-libraries/blob/develop/projects/rocprim/rocprim/include/rocprim/warp/detail/warp_reduce_shuffle.hpp#L72)) because the original CK implementation performs summation sequentially. Additionally, in the VLLM implementation, RMSnorm is done in a T5-like manner, where an intermediate value needs to be cast to BF16 (`temp = bf16(residual_v * s_variance)`). We adopted this approach to achieve similar value distribution, ultimately improving model accuracy (verified through the GSM8K test set).

However, we cannot guarantee the impact of these changes on all end-to-end applications. Therefore, we added a new pipeline pass (`rmsnorm2d_fwd_pipeline_model_sensitive_pass`) that allows users to choose between the new implementation and the original one. Users can leverage the framework to pass the `use_model_sensitive_rmsnorm` value via `rmsnorm2d_fwd_traits` to determine which specific RMS normalization implementation to use (currently, only the T5-like RMSnorm is added).

Additionally, this PR addresses the use of `rsqrtf` in the original RMSnorm pipeline to improve its precision.









## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [ ] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [x] Any dependent changes have been merged

## Discussion

If this is a relatively large or complex change, feel free to start a discussion by explaining why you chose the solution you did and what alternatives you considered

MHYangAMD committed:
Align casting to improve rmsnorm precision in BF16

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_pipeline_one_pass.hpp:
```diff
@@ -163,7 +184,19 @@ struct Rmsnorm2dFwdPipelineOnePass
-            Epilogue{}(y_window_, rmsn);
+            if constexpr(std::is_same_v<YDataType, ck_tile::bf16_t>)
+            {
+                store_tile(y_window_,
```
It would be better to put this code in Epilogue and let Epilogue take a new template parameter to choose the round mode.

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_pipeline_one_pass.hpp:
```diff
@@ -101,10 +101,36 @@ struct Rmsnorm2dFwdPipelineOnePass
-                store_tile(y_residual_window, cast_tile<YResidualDataType>(acc));
+                if constexpr(std::is_same_v<YResidualDataType, ck_tile::bf16_t>)
+                {
+                    store_tile(y_residual_window,
```
change `cast_tile` and let it take a new template parameter which defines rounding mode and has a default value for forward compatbility is better.

ruanjm commented on example/ck_tile/10_rmsnorm2d/generate.py:
```diff
@@ -84,7 +84,7 @@ class rmsnorm_fwd_codegen:
         if constexpr(is_warp_per_row)
         {
             static_assert(warpSize % ThreadPerBlock_N_ == 0);
-            return total_warps * (warpSize / ThreadPerBlock_N_);
```
Does this code really affect precision? If no, please remove it.

ruanjm commented on example/ck_tile/02_layernorm2d/generate.py:
```diff
@@ -85,7 +85,7 @@ class layernorm_fwd_codegen:
         if constexpr(is_warp_per_row)
         {
             static_assert(warpSize % ThreadPerBlock_N_ == 0);
-            return total_warps * (warpSize / ThreadPerBlock_N_);
```
Does this code really affect precision? If no, please remove it.

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_pipeline_one_pass.hpp:
```diff
@@ -132,9 +155,7 @@ struct Rmsnorm2dFwdPipelineOnePass
-
-            auto rmsn_ = acc[idx] * inv_rms_[i_idx] * gamma_;
-
+            const auto rmsn_  = acc[idx] * inv_rms_[i_idx] * gamma_;
```
Does this change really affect precision? If no, please remove it.

ruanjm commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -269,6 +269,37 @@ struct BlockReduce2dCrossWarpSync
+                    }
+                }
+                // Lane 0 now has the final reduced value for element i
+                if(lane_id == 0)
```
I'm thinking that __shfl_xor + removing `if(lane_id < num_reduce_warps)` + `smem_ptr[lane_id + i * num_warps]` -> `smem_ptr[lane_id % num_reduce_warps + i * num_warps]` should be able to let all the lanes to do the same thing. So that broadcast is no longer required.

MHYangAMD committed:
Workaround for performing tree reduction to follow vLLM

ClementLinCF committed:
Add Rmsnorm2dFwdPipelineModelSensitiveT5Pass

MHYangAMD committed:
Update rmsnorm2d_fwd_pipeline_model_sensitive_pass

1. Recover rmsnorm2d_fwd_pipeline_one_pass
2. Remove redundant casting for rmsnorm2d_fwd_pipeline_model_sensitive_pass
3. Add BlockReduce2dTreeCrossWarpSync

MHYangAMD committed:
Fix two pass cases

ClementLinCF committed:
Add Rmsnorm2dFusedModelSensitiveEnum

ClementLinCF committed:
fix clang-format

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_traits.hpp:
```diff
@@ -37,12 +37,24 @@ template<> struct Rmsnorm2dFusedQuantEnumName<Rmsnorm2dFusedQuantEnum::DYNAMIC_Q
+
+// clang-format off
+template<Rmsnorm2dFusedModelSensitiveEnum> struct Rmsnorm2dFusedModelSensitiveEnumName;
+template<> struct Rmsnorm2dFusedModelSensitiveEnumName<Rmsnorm2dFusedModelSensitiveEnum::T5_LIKED> { static constexpr const char * name = "t5"; };
```
Does `NO_MODEL_SENSITIVE` need a name as well?

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_traits.hpp:
```diff
@@ -37,12 +37,24 @@ template<> struct Rmsnorm2dFusedQuantEnumName<Rmsnorm2dFusedQuantEnum::DYNAMIC_Q
+enum class Rmsnorm2dFusedModelSensitiveEnum
+{
+    NO_MODEL_SENSITIVE = 0,
+    T5_LIKED           = 1,
```
Please provide more details about what `T5` is in comments.

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_traits.hpp:
```diff
@@ -37,12 +37,24 @@ template<> struct Rmsnorm2dFusedQuantEnumName<Rmsnorm2dFusedQuantEnum::DYNAMIC_Q
 template<> struct Rmsnorm2dFusedQuantEnumName<Rmsnorm2dFusedQuantEnum::SMOOTH_DYNAMIC_QUANT> { static constexpr const char * name = "smdqt"; };
 // clang-format on
 
+enum class Rmsnorm2dFusedModelSensitiveEnum
```
`Fused` in the name can be removed.

ruanjm commented on example/ck_tile/02_layernorm2d/generate.py:
```diff
@@ -85,7 +85,7 @@ class layernorm_fwd_codegen:
         {
             static_assert(warpSize % ThreadPerBlock_N_ == 0);
-            return total_warps * (warpSize / ThreadPerBlock_N_);
+            return total_warps;
```
Does this change really matter? Are you trying to change thread grid shape?

MHYangAMD replied:
Revert this.

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_pipeline_model_sensitive_pass.hpp:
```diff
@@ -0,0 +1,210 @@
+
+namespace ck_tile {
+
+template <typename Problem_, typename Policy_ = Rmsnorm2dFwdPipelineDefaultPolicy>
+struct Rmsnorm2dFwdPipelineModelSensitiveT5Pass
+{
```
It would be better to add some comments here to introduce this new pass. E.g. why it is named as T5, what the difference between it and original passes, whether this new pass is a variant either from one pass or two pass.

ruanjm commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -272,4 +272,126 @@ struct BlockReduce2dCrossWarpSync
+                smem_ptr[smem_offset + i * num_warps] = y_tensor.get_thread_buffer()[i];
+            });
+        }
+        block_sync_lds();
+
+        if(warp_id == 0)
+        {
+            static_for<0, thread_buf_size, 1>{}([&](auto i) {
+                DataType v;
+                if(lane_id < num_reduce_warps)
+                {
+                    v = smem_ptr[lane_id + i * num_warps];
+                }
+                else
+                {
+                    v = 0;
+                }
+                // Perform warp-wide reduction using shuffle operations
+                for(index_t offset = 1; offset < num_reduce_warps; offset <<= 1)
+                {
+                    // Each lane obtains the value from lane (lane_id + offset) of warp 0
+                    const index_t src_lane = lane_id + offset;
+                    DataType o             = warp_shuffle_down(v, offset);
+                    if(src_lane < num_reduce_warps)
+                    {
+                        v = reduce_func(v, o);
+                    }
+                }
+
+                if(lane_id == 0)
+                {
+                    smem_ptr[i * num_warps] = v;
+                }
+            });
+        }
+        block_sync_lds();
```
If let all the warp rather than only warp 0 to do these things, we should be able to remove w/r lds for broadcasting and the 2nd `block_sync_lds()`. I think this design is better for performance.

ruanjm commented on include/ck_tile/ops/reduce/block/block_reduce2d.hpp:
```diff
@@ -272,4 +272,126 @@ struct BlockReduce2dCrossWarpSync
+                    DataType o             = warp_shuffle_down(v, offset);
+                    if(src_lane < num_reduce_warps)
+                    {
+                        v = reduce_func(v, o);
```
You can use `__shfl_xor_sync` here rather than shuffle down because all the thread need the result finally.

ruanjm replied:
More details can be found in Shuffle Warp Reduce section of https://developer.nvidia.com/blog/faster-parallel-reductions-kepler/

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_pipeline_default_policy.hpp:
```diff
@@ -69,6 +69,15 @@ struct Rmsnorm2dFwdPipelineDefaultPolicy
         return BlockReduce2dCrossWarpSync<P_>{};
     }
 
+    template <typename Problem>
+    CK_TILE_HOST_DEVICE static constexpr auto GetBlockReduce2dTreeCrossWarpSync()
+    {
+        using P_ = BlockReduce2dProblem<typename Problem::ComputeDataType,
+                                        typename Problem::ComputeDataType,
+                                        typename Problem::BlockShape>;
+        return BlockReduce2dTreeCrossWarpSync<P_>{};
+    }
```
We can replace the previous cross warp sync with the tree version even for original passes if there is no performance regression.

ruanjm replied:
It is good to me to move this change to furture PR.

MHYangAMD replied:
Would leave this in future PR

ruanjm commented on example/ck_tile/10_rmsnorm2d/generate.py:
```diff
@@ -213,7 +218,19 @@ class rmsnorm_fwd_codegen:
+    using Pipeline = std::conditional_t<
+        (Traits_::kUseModelSensitiveRMSNorm == 0 || Traits_::kTwoPass), // TODO: consider TwoPass for T5PassPipeline
+        std::conditional_t<Traits_::kTwoPass, TwoPassPipeline, OnePassPipeline>,  // kUseModelSensitiveRMSNorm == 0
+        // std::conditional_t<
+        //     (Traits_::kUseModelSensitiveRMSNorm == 1),
+        //     T5PassPipeline,  // kUseModelSensitiveRMSNorm == 1
+        //     OtherPipeline    // kUseModelSensitiveRMSNorm == 2 (or other value)
+        // >
```
Remove the comments.

ruanjm commented on include/ck_tile/ops/rmsnorm2d/pipeline/rmsnorm2d_fwd_pipeline_model_sensitive_pass.hpp:
```diff
@@ -0,0 +1,210 @@
+        set_tile(square_sum, 0);
+        if constexpr(Problem::BlockShape::Vector_N % 2 == 0)
+        {
+            sweep_tile(
+                acc,
+                [&](auto idx_0, auto idx_1) {
+                    square_sum(idx_0) += acc[idx_0] * acc[idx_0] + acc[idx_1] * acc[idx_1];
+                },
+                sequence<1, 2>{});
```
I think it is better than the algorithm in `BlockReduce2d` in overwhelming majority of cases. It would be better to refactor `BlockReduce2d` with your new idea rather than do the special treatment here only.

You also need to do something special for vector with odd length.

ruanjm replied:
It is good to me to move this change to furture PR.

MHYangAMD replied:
Would leave this in future PR.

MHYangAMD committed:
Update patch

1. Reverse generate.py
2. Remove comment in generate.py
3. Update tree cross warp reduce

MHYangAMD committed:
Replace cross warp sync

ClementLinCF committed:
Refactor RMSNorm model enum and introduce T5-like option

ClementLinCF committed:
Merge branch 'develop' into ck_tile/rmsnorm-issue

MHYangAMD committed:
Update patch

1. Revert default cross warp policy
2. Update the n stage for cross warp reduce

MHYangAMD committed:
Merge branch 'develop' into ck_tile/rmsnorm-issue

ruanjm commented:
LGTM. Need code owner's approve.

MHYangAMD commented:
@carlushuang please help to approve this.

ClementLinCF committed:
Add new cmdline option in RMSNorm for new pipeline testing

ClementLinCF committed:
Merge branch 'develop' into ck_tile/rmsnorm-issue

carlushuang reviewed (approved):
LGTM
